### PR TITLE
KOGITO-3195: Fork KIE Editors

### PR DESCRIPTION
**Pull request by tiagobento**
We finally have a snapshot version of the new forked editors from `kogito-editors-java`. This PR is a huge milestone for Kogito Tooling, so it deserves a good description. I'll try to summarise everything included on this PR, as it's the result of a couple of month's work and we did many improvements that were not planned beforehand!

1. Update Editors to `8.0.0-SNAPSHOT-20210430.130828`
2. Remove `editorContext` from the `init` method of the Envelopes. This context is now part of the Init args we pass when opening Editors.
3. With the Ace editor removed from the forked Editors, we needed a new way to handle invalid files, so we added Monaco to the Online and Desktop channels, so there are some new UIs there.
4. Made Chrome Extension's token text input be closed by default.
5. Manually tree-shook `@patternfly/*` dependencies. When we're using ESM, this will be done automatically, but for now, we have to be extra-careful with our imports to not bring the entire PatternFly parts into our bundles.
6. Made the `Editor` interface accept custom APIs. This will be very useful when we need to provide different APIs for some of our Editors. Note that the Channels are still using the Editors with the default `Editor` API.
7. Added a custom action upon undo/redoing stuff on the `StateControl` class.
8. Removed `CompositeEditorFactory` as we don't want to Editors to be bundled in the same Envelope.
9. Improved CSS loading on the Online Editor. Please note that further improvements will be done on KOGITO-3560.

### CODE REVIEW

**Comment by tiagobento** [comment]
Important note: after this PR, the FDB build will be removed and the tests will be executed directly on kogito-editors-java in a new workflow. So please do not consider FDB on this PR.

**Comment by tiagobento** [comment]
@manstis Can you please take a look at the PMML Editor? I've changed a few things there so it would be great if you could check them! :)

**Review by caponetto — APPROVED** [review]
🚀

File: package.json [hunk 1]
```diff
@@ -22,10 +22,11 @@
     "format:check": "prettier --check ."
   },
   "dependencies": {
-    "@kogito-tooling/bpmn-editor-unpacked": "7.53.0-Final",
-    "@kogito-tooling/dmn-editor-unpacked": "7.53.0-Final",
+    "@kogito-tooling/bpmn-editor-unpacked": "8.0.0-SNAPSHOT-20210430.130828",
```
Comment (domhanak) [inline package.json:25]: I think it makes sense to consider using tags for editors version on our master e.g. "latest". I understand we already have a daily build&publish so just tagging the published version properly should do the trick. 

If we do not update this version for even a single day we risk that our tests will be yielding false positive results for our editors until we update the version again. With use of tag we can limit the time this happens to very small window, not ideal but far better imho.

Mentioning it here as I couldn't find more info in JIRAs and docs.

cc @Ginxo wdyt think about this approach of using tags?

File: packages/chrome-extension/src/app/components/pr/IsolatedPrEditor.tsx [hunk 2]
```diff
@@ -183,6 +183,9 @@ export function IsolatedPrEditor(props: {
           openFileExtension={props.fileExtension}
           readonly={true}
           keepRenderedEditorInTextMode={false}
+          onSetContentError={() => {
+            /**/
+          }} //FIXME: Tiago
```
Comment (tomasdavidorg) [inline packages/chrome-extension/src/app/components/pr/IsolatedPrEditor.tsx]: @tiagobento Could you please add a jira link here?

File: packages/desktop/static/envelope/bpmn-envelope.html [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      }
+    </style>
+
+    <title></title>
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
+  </head>
+  <body>
+    <div id="loading-screen" style="z-index: 100; position: relative"></div>
```
Comment (tomasdavidorg) [inline packages/desktop/static/envelope/bpmn-envelope.html]: I would say this line is not used. If this line and the line from `dmn-envelope.html` is removed, it will fix desktop it tests.
```suggestion
```

File: packages/desktop/static/envelope/dmn-envelope.html [hunk 4]
```diff
@@ -37,6 +37,6 @@
   <body>
     <div id="loading-screen" style="z-index: 100; position: relative"></div>
```
Comment (tomasdavidorg) [inline packages/desktop/static/envelope/dmn-envelope.html]: I would say this line is not used. If this line and the line from `bpmn-envelope.html` is removed, it will fix desktop it tests.
```suggestion
```

File: package.json (hunk 1 above)
Comment (tiagobento) [reply package.json:25]: Hi Dominik. This version is a snapshot from the kogito-editors-java, and this will not be included in any release. When we release a new version of the Editors, we run the sanity checks to know that everything is alright on the Kogito Tooling side. Also, there will be a new workflow on kogito-editors-java that will run the tests for kogito-tooling, so we guarantee on that repository that everything is alright for Kogito Tooling.

Using "tags" won't help, since they're concretized on our `yarn.lock` file upon installing. I've said that a couple of times to some people on the team, but there's nothing like Maven "SNAPSHOT" versions on NPM.

**Review by ederign — APPROVED** [review]

**Comment by ederign** [comment]
Great work Tiago, please take a look at the tests before merging it!

**Comment by tiagobento** [comment]
@tomasdavidorg @jstastny-cz @domhanak Hi guys! I'm getting Standalone Editors, Desktop, Online and Chrome Extension test failures on this PR, however, they're not failing consistently, as they pass on 2 out of 3 operating systems they're running on. Do you think we can assume it's random failures?

File: packages/chrome-extension/src/app/components/pr/IsolatedPrEditor.tsx (hunk 2 above)
Comment (tiagobento) [reply packages/chrome-extension/src/app/components/pr/IsolatedPrEditor.tsx]: Removed the FIXME and added a comment.

File: packages/desktop/static/envelope/bpmn-envelope.html (hunk 3 above)
Comment (tiagobento) [reply packages/desktop/static/envelope/bpmn-envelope.html]: Done. Thanks! Good catch.

File: packages/desktop/static/envelope/dmn-envelope.html (hunk 4 above)
Comment (tiagobento) [reply packages/desktop/static/envelope/dmn-envelope.html]: Done. Thanks! Good catch.

File: package.json (hunk 1 above)
Comment (domhanak) [reply package.json:25]: The new workflow sounds good, thanks for improvement!

I understand that this is snapshot version of the editors, thing is the moment something new gets in the editors repo our master is effectively out of date and any PR check that does not build the `kogito-editors-java` beforehand is invalid & could produce false positive.
Yes, as you mention, we have Full Downstream Build on `kogito-editors-java` and `kogito-tooling` which should mitigate this, yet in previous release we encountered a state where PR checks were green, but same test failed in nightly checks.
After some investigation. I found out it is because current IT tests do not use latest master.  @tiagobento I remember you mentioning this somewhere recently, but I am can't find where. Is there a JIRA? If not apologies.

The core of the issue is that not all of our PR checks are run against latest editors. 
* FDB - builds whole stack, does not run IT tests
* CI - builds kogito-tooling, runs only unit,snapshot tests
* IT tests (per channel) - build only kogito-tooling, runs only IT tests

I had a chat with @Ginxo he suggested to use `yarn link`  and modify our PR checks to always build the editor repository before the checks are run. It should happen for every check.  Is this something we can do? This would solve the issues we have and would allow us to keep master on latest released as we do now.
On top of that everytime we run tests with PR, we are sure they are run against the latest possible editors we have on master.

**Comment by domhanak** [comment]
@tiagobento I looked at standalone editors, however I have no mac to try and reproduce :( and video is not present, it seems we can't load a page with DMN and BPMN editors on mac-os. Did you encounter a green run on mac-os at least once? If yes I would say it is random, but it is worth investigating.

File: package.json (hunk 1 above)
Comment (Ginxo) [reply package.json:25]: In one hand I like the @domhanak proposal about using `latest`  `next` (or whatever tag we want to use)
In another hand I'm worried about cross-PR scenario where we will need to use not published npm libraries, in this case we will need to probably adapt the flows to build dependent projects and run yarn-link them before executing any test, wdyt?

**Comment by tomasdavidorg** [comment]
@tiagobento The desktop, Chrome and online editor tests are random failures. I ran all the tests locally in Windows virtual machine and they passed but I tried to lower the number of processors in the virtual machine to 2 and in some of the runs I was able to see this screen with two loading popups what is suspicious:
image

I saw this only when the online editor server was started and cypress tests were executed (processor usage was on 100%). Manually (without Cypress) I was not able to reproduce it.

File: package.json (hunk 1 above)
Comment (tiagobento) [reply package.json:25]: Sorry guys, as much as I find this topic very important and interesting, I think this is not related to this PR specifically. We already have a Jira planned to address that, so if you don't mind, let's move the discussion there: https://issues.redhat.com/browse/KOGITO-4860

**Comment by tiagobento** [comment]
@tomasdavidorg Thanks for taking a look! Chrome has a feature called CPU throttling that might be able to help reproduce that issue, have you tried it?

**Comment by tiagobento** [comment]
@domhanak Yes, Standalone Editors and Desktop are ok locally on my macOS.

**Comment by tomasdavidorg** [comment]
> @tomasdavidorg Thanks for taking a look! Chrome has a feature called CPU throttling that might be able to help reproduce that issue, have you tried it?

@tiagobento Thanks for the tip but I was still not able to reproduce it. With the The CPU throttling the loading popup is just there for a longer time and at the end the editor appears.

However I was able to reproduce it by opening 10 windows of online editor. On every I clicked "Try dmn sample" and 4 of them end up with the loading issue. In the console I found:
```
Uncaught (in promise) TypeError: Cannot read property 'languages' of undefined
    at uQk (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:24045)
    at vQk.wQk [as VC] (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:24343)
    at xQk.EQk [as VC] (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:24343)
    at dXk.aXk [as vD] (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:24344)
    at EXp.GXp [as KK] (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:24377)
    at e8o (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:7944)
    at c8o (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:19103)
    at Function.t8o (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:24372)
    at d (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:17236)
```

**Comment by tomasdavidorg** [comment]
I am able to reproduce one more issue. If I open 10 tabs with the online editor and in each I click "Create new decision model" I got this error in some of them: 
image

```
:9001/#/editor/dmn:1 Uncaught (in promise) TypeError: Cannot read property 'iR' of undefined
    at dXk.UWk [as dD] (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:24344)
    at nXk.vXk [as dD] (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:24344)
    at FVk.GVk [as dD] (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:24344)
    at RVk.VVk [as dD] (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:24344)
    at $wnd.GWTEditor.getContent (org.kie.workbench.common.dmn.showcase.DMNKogitoRuntimeWebapp-0.js:24012)
    at DmnEditorImpl.module.exports../src/common/GwtEditorWrapper.tsx.GwtEditorWrapper.getContent (index.js?7f42:351)
    at DmnEditorEnvelopeApiImpl.module.exports../src/envelope/KogitoEditorEnvelopeApiImpl.ts.KogitoEditorEnvelopeApiImpl.receive_contentRequest (index.js?a878:1140)
    at EnvelopeBusMessageManager.module.exports../src/common/EnvelopeBusMessageManager.ts.EnvelopeBusMessageManager.receive (index.js?69c4:345)
    at Object.receive (index.js?69c4:234)
    at EnvelopeBusController.module.exports../src/envelope/EnvelopeBusController.ts.EnvelopeBusController.receive (index.js?69c4:516)
    at eventListener (index.js?69c4:497)
```

File: package.json (hunk 1 above)
Comment (domhanak) [reply package.json:25]: @tiagobento Yes I agree. I ll move what we discussed into that jira.

**Comment by tiagobento** [comment]
@tomasdavidorg I can't reproduce those. Can you please give more details? Are you running it locally with `yarn start` on `packages/online-editor`? Are you using CPU throttling?

**Comment by tomasdavidorg** [comment]
@tiagobento I am able to reproduce it only on Windows 10 virtual machine (Virtual box). The machine has 2 CPUs, 7 GB RAM. Inside machine I start `yarn run start` from the online editor package. Then I open 10 tabs in Chrome browser with the online editor (http://localhost:9001/). I don't use CPU throttling.
- To reproduce the issues with 2 loading popups, I click `Try sample` (DMN) in all the tabs.
- To reproduce the issue with opening error, I click `Create new decision model"` in all the tabs.

https://user-images.githubusercontent.com/4498639/117101565-c81caa00-ad76-11eb-9277-0dc2efe95aa3.mp4

File: packages/pmml-editor/webpack.config.js [hunk 5]
```diff
@@ -31,7 +30,6 @@ module.exports = [
       libraryTarget: "commonjs2",
     },
     externals: [nodeExternals({ modulesDir: "../../node_modules" })],
-    plugins: [new MonacoWebpackPlugin()],
```
Comment (manstis) [inline packages/pmml-editor/webpack.config.js:34]: Now you've removed the `MonacoWebpackPlugin` the "standalone" PMML editor no longer has support for:-
- Syntax highlighting
- Code complete

You may need to add a similar entry to [this](https://github.com/kiegroup/kogito-tooling/pull/494/files#diff-68a6a2714cf5b5efe3d9a049201a3a04c9c3476dbdd3b25ab0398c57117f96e5R80) so that the standalone editor works correctly?

Comment (manstis) [reply packages/pmml-editor/webpack.config.js:34]: **`master` autocomplete (ctrl-space)**

master autocomplete

**`master` syntax highlighting**

master syntax highlighting

**`the_fork2` autocomplete missing (ctrl-space)**

No autocomplete

**`the_fork2` syntax highlighting missing**

No syntax highlighting

**Review by manstis — CHANGES_REQUESTED** [review]
The Monaco editor for "standalone" breaks with this PR.

Can you also advise why the need to change imports? Unless you communicate the reasoning to all contributors the old style will creep back in and whatever issue the change fixes will return.

**Comment by manstis** [comment]
Hi @kelvah I've added you for review too.

**Comment by jomarko** [comment]
### 01
Some monaco editor issue (https://github.com/kiegroup/kogito-tooling/pull/494#pullrequestreview-652133910) is present also for DMN in VS code.

Once autocompletion suggestions are displayed for DMN literal expression, the icons seems to be not loaded.

VS Code
Screenshot from 2021-05-05 16-09-21

dmn.new (online editor)
Screenshot from 2021-05-05 16-12-41

File: packages/pmml-editor/webpack.config.js (hunk 5 above)
Comment (tiagobento) [reply packages/pmml-editor/webpack.config.js:34]: Hi Michael!

WebpackMonacoPlugin was just a fancy way of rebuilding the Monaco Editor. We're actually fancier than that as we have our own publication of Monaco under `@kiegroup/monaco-editor` 😎

Jokes aside, I can't reproduce it. See this VSIX I just generared.
[vscode_extension_kogito_kie_editors_0.9.1.vsix.zip](https://github.com/kiegroup/kogito-tooling/files/6428036/vscode_extension_kogito_kie_editors_0.9.1.vsix.zip)

The icons are missing like @jomarko reported, though.

image
image

**Comment by tiagobento** [comment]
~Hi @jomarko thanks for reporting that. I really just don't wanna block this PR by this issue, if possible. Do you think we can create a separate Jira for this one? If so, we can use this Epic https://issues.redhat.com/browse/KOGITO-5039~

**EDIT**:
Fixed in 6663ddc7 for this repo. And here https://github.com/kiegroup/kogito-editors-java/pull/21 for kogito-editors-java.

Online Editor:
[online-editor-pr-494.zip](https://github.com/kiegroup/kogito-tooling/files/6429271/online-editor-pr-494.zip)

VSIX:
[vscode_extension_kogito_kie_editors_0.9.1.vsix.zip](https://github.com/kiegroup/kogito-tooling/files/6429279/vscode_extension_kogito_kie_editors_0.9.1.vsix.zip)

**Comment by tiagobento** [comment]
@tomasdavidorg Can you report this one under https://issues.redhat.com/browse/KOGITO-5039?

File: packages/pmml-editor/webpack.config.js (hunk 5 above)
Comment (manstis) [reply packages/pmml-editor/webpack.config.js:34]: @tiagobento The VSCode extension is not affected, only `pmml-editor` when run with `yarn run start`. This is the webapp we use use day to day development.

Comment (tiagobento) [reply packages/pmml-editor/webpack.config.js:34]: Oh, okay. Forgot about that! Good catch, I'll take a look.

Comment (tiagobento) [reply packages/pmml-editor/webpack.config.js:34]: Fixed in 1b458cf1!

**Review by jomarko — APPROVED** [review]
I can confirm the icons are now present id DMN visx.

- I tested manually the DMN and SCESIM editors. (Created small model with ~5 nodes and 4 scenarios for it). 
- I didn't checked code
- I didn't checked other platforms than VS Code.

**Comment by tomasdavidorg** [comment]
> @tomasdavidorg Can you report this one under https://issues.redhat.com/browse/KOGITO-5039?

@tiagobento Reported: https://issues.redhat.com/browse/KOGITO-5068

**Review by manstis — APPROVED** [review]
Thank-you for the fix @tiagobento.

**Comment by manstis** [comment]
@tiagobento I have approved; but please be sure to communicate "...why the need to change imports? Unless you communicate the reasoning to all contributors the old style will creep back in and whatever issue the change fixes will return".

**Comment by tiagobento** [comment]
@manstis The change on imports is related to item 5 on the description. Basically by importing the classes directly from their most specific directory, we are only importing what we really need. That's called tree-shaking. This is done automatically when using ES Modules, but since we're still on CommonJS, we had to do this change. There's an "automatic" way of doing it, but it's quite "magical" so I preferred to do that manually. 

On the upcoming weeks we'll be migrating from CommonJS to ESM, meaning that tree-shaking will be automatic and we will be able to import `@patternfly/react-core` as a whole without inflating our bundles unnecessarily. In the mean time, please use the most specific import path you can find.

**Comment by tiagobento** [comment]
@kelvah @domhanak @tomasdavidorg Already got approvals for QE and the PMML Editor team, can we proceed and merge this PR?

**Comment by tomasdavidorg** [comment]
@tiagobento Please wait for review by @LuboTerifaj. We agreed that @LuboTerifaj will do some more checks of the BPMN editor. Thanks.

**Review by tomasdavidorg — APPROVED** [review]

File: package.json [hunk 6]
```diff
@@ -22,10 +22,11 @@
...
+    "@kogito-tooling/dmn-editor-unpacked": "8.0.0-SNAPSHOT-20210430.130828",
     "@kogito-tooling/quarkus-runner-unpacked": "0.7.0-SNAPSHOT",
-    "@kogito-tooling/scesim-editor-unpacked": "7.53.0-Final",
+    "@kogito-tooling/scesim-editor-unpacked": "8.0.0-SNAPSHOT-20210430.130828",
+    "monaco-editor": "^0.23.0",
```
Comment (kelvah) [inline package.json:29]: I am bit confused. Why did you add a `monaco-editor` dep here? 
Maybe I'm wrong but isn't `@kiegroup/monaco-editor` acting like a wrapper for monaco? In that case I think it should replace it as a dependency

File: packages/pmml-editor/src/editor/PMMLEditor.tsx [hunk 7]
```diff
@@ -31,7 +31,7 @@ import { Provider } from "react-redux";
 import mergeReducers from "combine-reducer";
 import { HistoryContext, HistoryService } from "./history";
 import { LandingPage } from "./components/LandingPage/templates";
-import { Page } from "@patternfly/react-core";
+import { Page } from "@patternfly/react-core/dist/js/components/Page";
```
Comment (kelvah) [inline packages/pmml-editor/src/editor/PMMLEditor.tsx:34]: I saw your comments about "manual" tree shacking. Did you give a try to [what PF suggests](https://github.com/patternfly/patternfly-react/blob/master/ADVANCED-USAGE-README.md#tree-shaking) (`babel-plugin-transform-imports` for our case I think)?
Is it the magical solution you mentioned that wasn't ideal?
Asking because the manual change looks quite invasive.

**Review by kelvah — APPROVED** [review]
Hi @tiagobento, everything looks good for the pmml-editor. I left you a couple of questions. Thank you.

File: packages/pmml-editor/src/editor/PMMLEditor.tsx (hunk 7 above)
Comment (tiagobento) [reply packages/pmml-editor/src/editor/PMMLEditor.tsx:34]: Hi! Yes. I'm not saying it isn't ideal, I just didn't thought I wanted a tool dealing with the imports. It's one more thing to take care of and on the upcoming weeks we'll have ESM as default anyways..

File: package.json (hunk 6 above)
Comment (tiagobento) [reply package.json:29]: Good catch! `monaco-editor@^0.23.0` is listed as a devDependency on `@kiegroup/monaco-editor` as we didn't want to enforce the use of the original package when using it. The problem is that the type definitions are all in the original package, so we need to download it here too. When releasing `@kiegroup/monaco-editor@0.2.3` I didn't remember that, so we would need a new `@kiegroup/monaco-editor` release listing `monaco-editor@^0.23.0` as a dependency instead of devDependency. If it's okay I can do it on a new PR.

Comment (kelvah) [reply package.json:29]: It's okay for me :+1:

File: packages/pmml-editor/src/editor/PMMLEditor.tsx (hunk 7 above)
Comment (kelvah) [reply packages/pmml-editor/src/editor/PMMLEditor.tsx:34]: I understand. Thank you.

**Review by LuboTerifaj — APPROVED** [review]
Hi all,

I didn't find any issues in VSCode bpmn editor.

Thank you!

**Comment by tomasdavidorg** [comment]
Thanks @LuboTerifaj.

**Review by domhanak — APPROVED** [review]
